File: IPWorks Samples/CardDAV Client/c++/carddav.cpp
/*
 * IPWorks 2024 C++ Edition - Sample Project
 *
 * This sample project demonstrates the usage of IPWorks in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/ipworks
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */


#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <string>
#include <vector>
#include <iostream>
#include <iomanip> 
#include <sstream>
#include <chrono>

#include "../../include/ipworks.h"
#define LINE_LEN 100

using namespace std;

bool exportingContact = false;

void getAuthorization(CardDAV* carddav, OAuth* oauth, string password, string clientID, string clientSecret) {
  if (clientID.empty() && clientSecret.empty()) {
    carddav->SetPassword(password.c_str());
  }
  else {
    oauth->SetClientId(clientID.c_str());
    oauth->SetClientSecret(clientSecret.c_str());
    oauth->SetServerAuthURL("https://accounts.google.com/o/oauth2/auth");
    oauth->SetServerTokenURL("https://oauth2.googleapis.com/token");
    oauth->SetAuthorizationScope("https://www.googleapis.com/auth/carddav");
    carddav->SetAuthScheme(AUTH_OAUTH);
    carddav->SetAuthorization(oauth->GetAuthorization());
  }
}

class MyCardDAV : public CardDAV {
public:
  int contactNumber = 1;

  MyCardDAV() {
    contactNumber = 1;
  }

  virtual int FireContactDetails(CardDAVContactDetailsEventParams* e) {
    string firstPhoneNumber = " - ";
    if (this->GetPhoneNumbers()->GetCount() > 0) {
      firstPhoneNumber = this->GetPhoneNumbers()->Get(0)->GetValue();
    }
    string firstEmail = " - ";
    if (this->GetEMails()->GetCount() > 0) {
      firstEmail = this->GetEMails()->Get(0)->GetValue();
    }
    cout << left << contactNumber
                << setw(4) << ")"
                << setw(40) << this->GetFormattedName()
                << setw(30) << firstPhoneNumber
                << setw(30) << firstEmail
                << setw(30) << e->ResourceURI << "\n\n";
    contactNumber++;
    return 0;
  }

  virtual int FireSSLServerAuthentication(CardDAVSSLServerAuthenticationEventParams* e) {
    e->Accept = true;  //this will trust all certificates and it is not recommended for production use
    return 0;
  }

  virtual int FireLog(CardDAVLogEventParams* e) {
    //cout << e->Message;
    return 0;
  }

};

int main(int argc, char* argv[]) {

  string command;     // user's command
  string argument;             // arguments to the user's command	
  string clientID;
  string clientSecret;
  int ret_code = 0;
  MyCardDAV carddav;
  OAuth* oauth = new OAuth();

  cout << "********************************************************************************\n";
  cout << "* This demo shows how to use the CardDAV component to list contacts from an    *\n";
  cout << "* existing Google or NextCloud address book. You can also create a new         *\n";
  cout << "* contact, delete a contact, or export a contact to a vcf file.                *\n";
  cout << "********************************************************************************\n\n";

  if (argc < 4) {
    cerr << "usage: carddav [options] uri username password\n";
    cerr << "Options\n";
    cerr << "  -i        OAuth Client ID\n";
    cerr << "  -s        OAuth Client Secret\n";
    cerr << "  URI       the URI of the address book \n";
    cerr << "  username  the username to login\n";
    cerr << "  password  the password to login\n";
    cerr << "\nExample: carddav -i OAuthClientID -s OAuthClientSecret URI username password\n\n";
    cout << "Press enter to continue.";
  }
  else {
    cout << "\nType \"?\" for a list of commands.\n";

    while (true) {
      for (int i = 0; i < argc; i++) {

        if (strncmp(argv[i], "-", 1) == 0) { //check if argument starts with "-"
          if (strcmp(argv[i], "-i") == 0) {
            clientID = argv[i + 1];
          }
          if (strcmp(argv[i], "-s") == 0) {
            clientSecret = argv[i + 1];
          }
        }
      }

      exportingContact = false; //This is used so output from ContactDetails is not displayed when exporting an even to an ics file
      cout << "> ";
      cin >> command;
      istringstream iss(command);

      iss >> argument;

      string user = argv[argc - 2];
      string password = argv[argc - 1];
      string mainAddressbookURL = argv[argc - 3];

      if (argument.compare("?") == 0) {
        cout << "1) List Events\n";
        cout << "2) Add Event\n";
        cout << "3) Delete Event\n";
        cout << "4) Export ICS File\n";
        cout << "?) This help menu.\n";
        cout << "Q) Quit.\n";
      }
      else if (argument.compare("1") == 0) {  //List Events
        carddav.contactNumber = 1;
        cout << left << setfill(' ') << ""
          << setw(5) << ""
          << setw(40) << "Name"
          << setw(30) << "PhoneNumber"
          << setw(30) << "Email"
          << setw(30) << "ResourceURI" << "\n\n";
        carddav.SetUser(user.c_str());
        getAuthorization(&carddav, oauth, password, clientID, clientSecret);
        carddav.GetAddressbookReport(mainAddressbookURL.c_str());
        carddav.Reset();
      }
      else if (argument.compare("2") == 0) {    //Add Event
        carddav.SetUser(user.c_str());
        getAuthorization(&carddav, oauth, password, clientID, clientSecret); 

        auto now = chrono::system_clock::now();
        auto millis = chrono::time_point_cast<chrono::milliseconds>(now);
        carddav.SetUID(to_string(millis.time_since_epoch().count()).c_str());
        cout << "Name: ";
        string name;
        cin >> name;
        carddav.SetFormattedName(name.c_str());
        cout << "Phone number: ";
        string phoneNumber;
        cin >> phoneNumber;
        carddav.GetPhoneNumbers()->SetCount(1);
        carddav.SetPhoneNumberName(0, "TEL");
        carddav.SetPhoneNumberValue(0, phoneNumber.c_str());
        cout << "Email: ";
        string email;
        cin >> email;
        carddav.GetEMails()->SetCount(1);
        carddav.SetEMailName(0, "EMAIL");
        carddav.SetEMailValue(0, email.c_str());
        cout << "Address: ";
        string address;
        cin >> address;
        carddav.GetAddresses()->SetCount(1);
        carddav.SetAddressName(0, "ADR");
        carddav.SetAddressValue(0, address.c_str());
        carddav.Config("loglevel=3");
        carddav.CreateContact((mainAddressbookURL + "/" + carddav.GetUID() + ".vcf").c_str());
        cout << "Contact created successfully";
      }
      else if (argument.compare("3") == 0) {     //Delete Event
        cout << "Give name of file to delete (find the file names by running option #1): ";
        string filename;
        cin >> filename;
        carddav.SetUser(user.c_str());
        getAuthorization(&carddav, oauth, password, clientID, clientSecret);
        carddav.DeleteContact((mainAddressbookURL + "/" + filename).c_str());
        cout << "Contact successfully deleted\n";
      }
      else if (argument.compare("4") == 0) {     //Export Event
        cout << "Give name of the .vcf file (find the file names by running option #1): ";
        string filename;
        cin >> filename;
        carddav.SetUser(user.c_str());
        getAuthorization(&carddav, oauth, password, clientID, clientSecret);
        cout << left << setfill(' ') << ""
          << setw(5) << ""
          << setw(40) << "Name"
          << setw(30) << "PhoneNumber"
          << setw(30) << "Email"
          << setw(30) << "ResourceURI" << "\n\n";
        carddav.GetContact((mainAddressbookURL + "/" + filename).c_str());
      }
      else if (argument.compare("q") == 0 || argument.compare("Q") == 0) {
        exit(0);
        return 0;
      }
      else if (argument.empty()) {
        // Do nothing
      }
      else {
        cout << "Bad command / Not implemented in demo.\n";
      }
      if (ret_code)    // Got an error.
      {
        cout << "\nError: %d", ret_code;
        if (carddav.GetLastError()) {
          cout << " \"%s\"\n", carddav.GetLastError();
        }
      }
      ret_code = 0; //flush out error
    }
  }

  exit(ret_code);
  return 0;
}

File: IPWorks Samples/CardDAV Client/java/carddav.java
/*
 * IPWorks 2024 Java Edition - Sample Project
 *
 * This sample project demonstrates the usage of IPWorks in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/ipworks
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */

import java.io.*;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;

import ipworks.CardCustomProp;
import ipworks.CardDAV;
import ipworks.CardDAVContactDetailsEvent;
import ipworks.CardDAVSSLServerAuthenticationEvent;
import ipworks.DefaultCardDAVEventListener;
import ipworks.IPWorksException;
import ipworks.OAuth;

public class carddav extends ConsoleDemo {

    private static CardDAV carddav = new CardDAV();
    private static OAuth oauth = new OAuth();
    private static String[] argument;
    private static String clientID = "";
    private static String clientSecret = "";
    private static List<String> resourceURIs = new ArrayList<>();
    private static boolean exportingContact = false;
    private static int contactNumber = 1;
    private static String user;
    private static String password;

    public static void main(String[] args) {
        
        if (args.length < 3) {
            System.out.println("usage: carddav [options] URI username password");
            System.out.println("Options");
            System.out.println("  -i        OAuth Client ID");
            System.out.println("  -s        OAuth Client Secret");
            System.out.println("  URI       the URI of the address book");
            System.out.println("  username  the username to login");
            System.out.println("  password  the password to login");
            System.out.println("\r\nExample: carddav -i OAuthClientID -s OAuthClientSecret URI username password");
        } else {
            try {
                carddav.addCardDAVEventListener(new DefaultCardDAVEventListener() {
                    public void SSLServerAuthentication(CardDAVSSLServerAuthenticationEvent e) {
                        e.accept = true; //this will trust all certificates and it is not recommended for production use
                    }
                    public void contactDetails(CardDAVContactDetailsEvent e) {
                        if (!exportingContact) {
                            resourceURIs.add(e.resourceURI);
                            String phoneNumber = (carddav.getPhoneNumbers().size() > 0 && !carddav.getPhoneNumbers().get(0).getValue().trim().isEmpty()) ? carddav.getPhoneNumbers().get(0).getValue() : " - ";
                            String email = (carddav.getEMails().size() > 0 && !carddav.getEMails().get(0).getValue().trim().isEmpty()) ? carddav.getEMails().get(0).getValue() : " - ";
                            System.out.printf("%-5s %-40s %-30s %-30s %-30s%n", (contactNumber + ")"), carddav.getFormattedName(), phoneNumber, email, e.resourceURI);
                            contactNumber++;
                        }
                    }

                });
                System.out.println("********************************************************************************");
                System.out.println("* This demo shows how to use the CardDAV component to list contacts from an    *");
                System.out.println("* existing Google or NextCloud address book. You can also create a new         *");
                System.out.println("* contact, delete a contact, or export a contact to a vcf file.                *");
                System.out.println("********************************************************************************\n");
                System.out.println("Type \"?\" for a list of commands.");

                while (true) {
                    try {
                        exportingContact = false;
                        System.out.print("> ");
                        argument = input().split(" ");

                        for (int i = 0; i < args.length; i++) {
                            if (args[i].startsWith("-")) {
                                if (args[i].equals("-i")) {
                                    clientID = args[i + 1];
                                }
                                if (args[i].equals("-s")){
                                    clientSecret = args[i + 1];
                                } 
                            }
                        }

                        user = args[args.length - 2];
                        password = args[args.length - 1];
                        String mainAddressbookURL = args[args.length - 3];
                        
                        if (argument[0].equals("?")) {
                            System.out.println("1) List Contacts");
                            System.out.println("2) Add Contact");
                            System.out.println("3) Delete Contact");
                            System.out.println("4) Export VCF File");
                            System.out.println("?) Help menu.");
                            System.out.println("Q) Quit.");

                        } else if (argument[0].equals("1")) {

                            resourceURIs.clear();
                            contactNumber = 1;
                            System.out.printf("%-5s %-40s %-30s %-30s %-30s%n", "", "Name", "PhoneNumber", "Email", "ResourceURI\n");
                            getAuthorization();
                            carddav.getAddressbookReport(mainAddressbookURL);
                            carddav.reset();

                        } else if (argument[0].equals("2")) {

                            getAuthorization();
                            carddav.setUID(String.valueOf(System.currentTimeMillis()));
                            carddav.setFormattedName(prompt("Full Name"));
                            carddav.getPhoneNumbers().add(new CardCustomProp("TEL", prompt("Phone number")));
                            carddav.getEMails().add(new CardCustomProp("TEL", prompt("Email")));
                            carddav.getAddresses().add(new CardCustomProp("TEL", prompt("Address")));
                            carddav.createContact(mainAddressbookURL + "/" + carddav.getUID() + ".vcf");
                            System.out.println("Contact successfully added");

                        } else if (argument[0].equals("3")) {

                            String resource = prompt("Give name of file to delete (find the file names by running option #1)");
                            getAuthorization();
                            carddav.deleteContact(mainAddressbookURL + "/" + resource);
                            System.out.println("Contact successfully deleted");

                        } else if (argument[0].equals("4")) {

                            String resource = prompt("Give name of the .vcf file (find the file names by running option #1)");
                            getAuthorization();
                            System.out.printf("%-5s %-40s %-30s %-30s %-30s%n", "", "Name", "PhoneNumber", "Email", "ResourceURI\n");
                            carddav.getContact(mainAddressbookURL + "/" + resource);
                            //File.WriteAllText(resource, carddav.ExportVCF());
                        } else if (argument[0].toLowerCase().equals("q")) {

                            System.exit(0);
                            return;

                        } else if (argument[0].equals("")) {
                            // do nothing
                        } else {
                            System.out.println("Bad command / Not implemented in demo.");
                        }
                    } catch (IPWorksException e) {
                        System.out.println(e.getMessage());
                        System.exit(0);
                        return;
                    }
                }
            } catch (Exception e) {
                displayError(e);
            }
        }
    }

    private static void getAuthorization() throws IPWorksException {
        carddav.setUser(user);
        if(clientID.isEmpty() && clientSecret.isEmpty()) {
            carddav.setPassword(password);
        }
        else {  
            oauth.setClientId(clientID);
            oauth.setClientSecret(clientSecret);
            oauth.setServerAuthURL("https://accounts.google.com/o/oauth2/auth");
            oauth.setServerTokenURL("https://accounts.google.com/o/oauth2/token");
            oauth.setAuthorizationScope("https://www.googleapis.com/auth/carddav");
            carddav.setAuthScheme(CardDAV.authOAuth);
            carddav.setAuthorization(oauth.getAuthorization());
        }

    }

}
class ConsoleDemo {
  private static BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));

  static String input() {
    try {
      return bf.readLine();
    } catch (IOException ioe) {
      return "";
    }
  }
  static char read() {
    return input().charAt(0);
  }

  static String prompt(String label) {
    return prompt(label, ":");
  }
  static String prompt(String label, String punctuation) {
    System.out.print(label + punctuation + " ");
    return input();
  }
  static String prompt(String label, String punctuation, String defaultVal) {
      System.out.print(label + " [" + defaultVal + "] " + punctuation + " ");
      String response = input();
      if (response.equals(""))
        return defaultVal;
      else
        return response;
  }

  static char ask(String label) {
    return ask(label, "?");
  }
  static char ask(String label, String punctuation) {
    return ask(label, punctuation, "(y/n)");
  }
  static char ask(String label, String punctuation, String answers) {
    System.out.print(label + punctuation + " " + answers + " ");
    return Character.toLowerCase(read());
  }

  static void displayError(Exception e) {
    System.out.print("Error");
    if (e instanceof IPWorksException) {
      System.out.print(" (" + ((IPWorksException) e).getCode() + ")");
    }
    System.out.println(": " + e.getMessage());
    e.printStackTrace();
  }
}



